Description: Change location of license file in about dialog
Forwarded: not-needed
Author: Frederic Peters <fpeters@debian.org>

--- a/ui/qt/about_dialog.cpp
+++ b/ui/qt/about_dialog.cpp
@@ -335,7 +335,7 @@
 #if defined(_WIN32)
     f_license.setFileName(get_datafile_path("COPYING.txt"));
 #else
-    f_license.setFileName(get_datafile_path("COPYING"));
+    f_license.setFileName(get_datafile_path("ABOUT.GPL"));
 #endif

     f_license.open(QFile::ReadOnly | QFile::Text);
