Description: Change location of license file in about dialog
Forwarded: not-needed
Author: Frederic Peters <fpeters@debian.org>

--- a/ui/gtk/about_dlg.c
+++ b/ui/gtk/about_dlg.c
@@ -506,7 +506,7 @@
 #if defined(_WIN32)
   absolute_path = get_datafile_path("COPYING.txt");
 #else
-  absolute_path = get_datafile_path("COPYING");
+  absolute_path = get_datafile_path("ABOUT.GPL");
 #endif
   page = text_page_new(absolute_path);
