--- glib-2.29.8/glib/gconvert.c.orig	2011-06-26 14:48:37.000000000 +0200
+++ glib-2.29.8/glib/gconvert.c	2011-06-26 14:49:58.000000000 +0200
@@ -61,7 +61,7 @@
 #if defined(USE_LIBICONV_GNU) && !defined (_LIBICONV_H)
 #error GNU libiconv in use but included iconv.h not from libiconv
 #endif
-#if !defined(USE_LIBICONV_GNU) && defined (_LIBICONV_H)
+#if !(defined(__APPLE__) && defined(__LP64__)) && !defined(USE_LIBICONV_GNU) && defined (_LIBICONV_H)
 #error GNU libiconv not in use but included iconv.h is from libiconv
 #endif
