--- src/init.c.orig	2021-03-15 00:11:33
+++ src/init.c	2023-06-06 03:15:37
@@ -269,7 +269,7 @@
  *
  * @see ssh_init()
  */
-bool is_ssh_initialized() {
+bool is_ssh_initialized(void) {
 
     bool is_initialized = false;
 
--- examples/sshnetcat.c.orig	2021-08-26 05:27:42
+++ examples/sshnetcat.c	2023-06-06 03:20:21
@@ -233,7 +233,7 @@
 }
 
 void cleanup_pcap(void);
-void cleanup_pcap(){
+void cleanup_pcap(void){
 	ssh_pcap_file_free(pcap);
 	pcap=NULL;
 }
