--- snappy.cc.orig	2023-09-14 01:04:05
+++ snappy.cc	2023-09-14 01:04:28
@@ -1290,7 +1290,7 @@
         DeferMemCopy(&deferred_src, &deferred_length, from, len);
       }
     } while (ip < ip_limit_min_slop &&
-             (op + deferred_length) < op_limit_min_slop);
+             static_cast<ptrdiff_t>(op + deferred_length) < op_limit_min_slop);
   exit:
     ip--;
     assert(ip <= ip_limit);
